From 75bcb0328db8ef1e274158104975aeb942639c5f Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Fri, 11 Sep 2020 23:36:46 +0300
Subject: [PATCH] [SW-17444] preboot: goya2_arc: add svn validation

This patch adds support to verify minimum allowed security version
number as part of secure boot flow.

As part of ARC based secure boot flow, arcboot collects all the
information such as svn, revoke bits and key hash from efuse, otp
or flash and copies it at start of scratchpad SRAM. preboot while
performing image verification re-use the above information stored
in scratchpad SRAM by using proper offset shared between arcboot
and preboot

Change-Id: I34e18805d35d840803b19383b1ba2b0772505df3
---

diff --git a/goya2_arc_gnu/Makefile b/goya2_arc_gnu/Makefile
index c142d30..31c01e0 100644
--- a/goya2_arc_gnu/Makefile
+++ b/goya2_arc_gnu/Makefile
@@ -2,3 +2,4 @@
 
 PREBOOT_SRCS_C += $(shell ls $(HL_DEV)/pll.c)
 PREBOOT_SRCS_C += $(shell ls preboot/thermal_v2.c)
+PREBOOT_SRCS_C += $(shell ls $(HL_DEV)/security.c)
diff --git a/goya2_arc_gnu/security.c b/goya2_arc_gnu/security.c
new file mode 100644
index 0000000..c9b8564
--- /dev/null
+++ b/goya2_arc_gnu/security.c
@@ -0,0 +1,33 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Copyright 2020 HabanaLabs, Ltd.
+ * All Rights Reserved.
+ *
+ */
+
+#include <preboot.h>
+#include <string.h>
+#include <uart.h>
+#include <preboot/sha256.h>
+#include <goya2/goya2_security.h>
+
+static u32 sp_sram_readl(u32 offset)
+{
+	u32 reg = *((volatile u32 *) (HL_SCRATCHPAD_SRAM_BASE + offset));
+	return reg;
+}
+
+bool preboot_security_svn_is_valid(int sec_ver)
+{
+	u8 sec_fw_ver, sec_valid_fw_ver;
+	u32 valid_fw_ver;
+
+	sec_fw_ver = SECURITY_FW_VERSION_TO_SECURITY_BYTE(sec_ver);
+	valid_fw_ver = sp_sram_readl(SECURITY_EFUSE_FW_VERSION_OFFSET);
+	sec_valid_fw_ver = SECURITY_FW_VERSION_TO_SECURITY_BYTE(valid_fw_ver);
+
+	if (sec_fw_ver < sec_valid_fw_ver)
+		preboot_puts_hex("Security version too low: ", sec_ver, 0);
+
+	return sec_fw_ver >= sec_valid_fw_ver;
+}
